### PR TITLE
feat: prune versions without bookkeepings

### PR DESCRIPTION
**Pull request by yihuang**

### CODE REVIEW

File: iavl/diff.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        yield l1.nodes, []
+        l1.next_layer(ndb)
+
+    while l2.height > l1.height:
+        yield [], l2.nodes
+        l2.next_layer(ndb)
+
+    while True:
+        # l1 l2 at the same height now
+        common, orphaned, new = diff_sorted(l1.nodes, l2.nodes)
```
Comment (mmsqe) [inline iavl/diff.py:111]: may I ask the usage of this `common`?

File: iavl/diff.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        yield orphaned, new
+
+        if l1.height == 0:
+            break
+
+        # don't visit the common sub-trees
+        l1.nodes = orphaned
+        l2.nodes = new
```
Comment (mmsqe) [inline iavl/diff.py:119]: oh so when pass l1 as 1st param, orphaned means exist in l1 but no exist in l2, while new means not exist in l1 but exist in l2, then remove the common nodes from both layers by reassign

Comment (yihuang) [reply iavl/diff.py:119]: exactly👍

File: iavl/diff.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+"""
+from .iavl import NodeDB
+
+
+class Layer:
+    """
+    Represent one layer of nodes at the same height
+
+    pending_nodes: because one of the children's height could be height-2, need to keep
+    it in the pending list temporariliy.
```
Comment (yihuang) [inline iavl/diff.py:12]:
```suggestion
    it in the pending list temporarily.
```

File: iavl/diff.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        # merge sorted lists
+        self.nodes = nodes
+        self.nodes += self.pending_nodes
+        self.nodes.sort(key=lambda n: n.key)
+        self.pending_nodes = pending_nodes
+
+
+def diff_sorted(nodes1, nodes2):
+    """
+    return (common, orphaned, new)
```
Comment (yihuang) [inline iavl/diff.py:61]:
```suggestion
    Contract: input list is sorted by node.key
    return: (common, orphaned, new)
```

Comment (mmsqe) [reply iavl/diff.py]: what's this Contract?

File: iavl/diff.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        if i2 > len(nodes2) - 1:
+            orphaned += nodes1[i1:]
+            break
+        k1 = nodes1[i1].key
+        k2 = nodes2[i2].key
+        if nodes1[i1].hash == nodes2[i2].hash:
+            common.append(nodes1[i1])
+            i1 += 1
+            i2 += 1
+        elif k1 == k2:
```
Comment (mmsqe) [inline iavl/diff.py:80]:
```suggestion
        elif k1 == k2:
        # overriden by same key
```

File: iavl/diff.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            orphaned.append(nodes1[i1])
+            new.append(nodes2[i2])
+            i1 += 1
+            i2 += 1
+        elif k1 < k2:
```
Comment (mmsqe) [inline iavl/diff.py:85]:
```suggestion
        elif k1 < k2:
        # proceed to next node in nodes1 until same key with nodes2
```

File: iavl/diff.py (hunk 1 above)
Comment (yihuang) [reply iavl/diff.py:115]: the common is not used here actually.
```suggestion
        _, orphaned, new = diff_sorted(l1.nodes, l2.nodes)
```

File: iavl/cli.py [hunk 7]
```diff
@@ -344,5 +356,44 @@ def visualize(db, version, store=None, include_prev_version=False):
...
+@click.option(
+    "--db", help="path to application.db", type=click.Path(exists=True), required=True
+)
+@click.option("--store", "-s")
+@click.option(
+    "--version",
+    help="the version to query, default to latest version if not provided",
+    type=click.INT,
+)
+def state_changes(db, version, store: Optional[str]):
```
Comment (mmsqe) [inline iavl/cli.py:369]: same [reason](https://github.com/crypto-com/python-iavl/pull/3#discussion_r1034392113) for optional store right?

File: iavl/diff.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+    while l2.height > l1.height:
+        yield [], l2.nodes
+        l2.next_layer(get_node)
+
+    while True:
+        # l1 l2 at the same height now
+        _, orphaned, new = diff_sorted(l1.nodes, l2.nodes)
+
+        if not orphaned and not new:
```
Comment (mmsqe) [inline iavl/diff.py:129]: so no need handle empty when layer unchanged right

File: iavl/cli.py [hunk 9]
```diff
@@ -344,5 +356,44 @@ def visualize(db, version, store=None, include_prev_version=False):
...
+    db = dbm.open(str(db), read_only=True)
+    if version is None:
+        version = iavl_latest_version(db, store)
+
+    prefix = store_prefix(store) if store is not None else b""
+
+    def get_node(hash: bytes) -> Optional[Node]:
```
Comment (mmsqe) [inline iavl/cli.py:376]: might mv `get_node` to utils

File: iavl/cli.py (hunk 7 above)
Comment (yihuang) [reply iavl/cli.py:369]: yes

File: iavl/cli.py (hunk 9 above)
Comment (yihuang) [reply iavl/cli.py:376]: it use the prefix from the closure.

Comment (mmsqe) [reply iavl/cli.py:376]: oh right, need pass more params

**Review by mmsqe — APPROVED** [review]

File: iavl/diff.py (hunk 8 above)
Comment (yihuang) [reply iavl/diff.py:129]: yes, won't yield empty list.

File: iavl/cli.py (hunk 9 above)
Comment (yihuang) [reply iavl/cli.py]: refactored, much better now.

File: iavl/diff.py (hunk 8 above)
Comment (yihuang) [reply iavl/diff.py]: this is actually buggy, because there's pending nodes.

File: iavl/cli.py [hunk 10]
```diff
@@ -344,5 +343,82 @@ def visualize(db, version, store=None, include_prev_version=False):
...
+@click.option(
+    "--db", help="path to application.db", type=click.Path(exists=True), required=True
+)
+@click.option("--store", "-s")
+@click.option(
+    "--start-version",
+    help="the version to start check",
+    default=1,
+)
+def test_state_round_trip(db, store, start_version):
```
Comment (yihuang) [inline iavl/cli.py:386]: running this cmd on testnet db, so far so good.

File: iavl/diff.py [hunk 11]
```diff
@@ -196,7 +193,14 @@ def state_changes(get_node: GetNode, root1: PersistedNode, root2: PersistedNode)
     """
     for orphaned, new in diff_tree(get_node, root1, root2):
         # the nodes are on the same height, and we only care about leaf nodes here
-        node = orphaned[0] if orphaned else new[0]
+        try:
```
Comment (mmsqe) [inline iavl/diff.py:196]: since we cant handle new[0] index out bound before?

File: iavl/diff.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    """
+    for orphaned, new in diff_tree(get_node, root1, root2):
+        # the nodes are on the same height, and we only care about leaf nodes here
+        try:
+            node = orphaned[0]
+        except IndexError:
+            try:
+                node = new[0]
+            except IndexError:
+                continue
```
Comment (mmsqe) [inline iavl/diff.py:202]:
```suggestion
        node = orphaned[0] if len(orphaned) > 0 else (new[0] if len(new) > 0 else None)
        if node is None:
            continue
```

**Pull request merged**